polartools/diffraction.py

```diff
@@ -582,10 +585,177 @@ def load_series(
+        scan_type = scan_cmd[0]
+        if scan_type == "mesh" or scan_type == "hklmesh":
+            scan_range = {
+                "x0": scan_cmd[2],
+                "x1": scan_cmd[3],
+                "xint": scan_cmd[4],
+                "y0": scan_cmd[6],
+                "y1": scan_cmd[7],
+                "yint": scan_cmd[8],
+            }
+        elif scan_type != "qxscan":
+            scan_range = {
+                "x0": scan_cmd[2],
+                "x1": scan_cmd[3],
+                "xint": scan_cmd[4],
+            }
+        else:
+            scan_range = {}
+    else:
+        scan_type = collect_meta(scan_id, source, ["plan_name"])
+    return scan_type, scan_range
+
+
+def load_mesh(scan, source, scan_range, **kwargs):
+    """
+    Load mesh generates input array for plot_2d from mesh_scan.
+
+    Parameters
+    ----------
+    source : databroker database, name of the spec file, or 'csv'
+        Note that applicable kwargs depend on this selection.
+    scan_series : list, int
+        start, stop, step, [start2, stop2, step2, ... ,startn, stopn, stepn]
+    scan_range : list, int
+        scan parameters of mesh scan [x0, x1, xinterval, y0, y1, yinterval]
+
+    kwargs :
+        The necessary kwargs are passed to the loading and fitting functions
+        defined by the `source` argument:
+            - csv        -> possible kwargs: folder, name_format, e.g.\
+                "scan_{}_primary.csv"
+            - spec       -> possible kwargs: folder
+            - databroker -> possible kwargs: stream, query
+
+        Note that a warning will be printed if the an unnecessary kwarg is
+        passed.
+
+    Additional parameters in kwargs:
+        scale : list, int
+            intensity limits: [z_min,z_max]
+
+    Returns
+    -------
+    data : arrays with x, y and z information for 2D plot and axes names
+    """
+
+    folder = kwargs.pop("folder", "")
+    scale = kwargs.pop("scale", [None, None])
+    if source == "csv":
+        pass
+        # to be implemented for csv
+    elif isinstance(source, str) or isinstance(source, SpecDataFile):
+        if isinstance(source, str):
+            path = join(folder, source)
+            source = SpecDataFile(path)
+        specscan = source.getScan(scan)
+
+    data = load_table(scan=scan, source=source)
+    x_label = data.columns[0]
+    y_label = data.columns[1]
+    z_label = data.columns[-1]
+
+    x1 = [
+        x1[:]
+        for x1 in [[1.01] * (int(scan_range["xint"]) + 1)]
+        * (int(scan_range["yint"]) + 1)
+    ]
+    y1 = [
+        y1[:]
+        for y1 in [[1.01] * (int(scan_range["xint"]) + 1)]
+        * (int(scan_range["yint"]) + 1)
+    ]
+    z1 = [
+        z1[:]
+        for z1 in [[1.01] * (int(scan_range["xint"]) + 1)]
+        * (int(scan_range["yint"]) + 1)
+    ]
+
+    r0 = specscan.data[x_label]
+    r1 = specscan.data[y_label]
+    r2 = specscan.data[z_label]
+
+    for ii in range(0, int(scan_range["yint"]) + 1):
```

Comment:
Can this be done using matrix operations? Generally speaking, it is **much** more efficient to operate on `numpy.array` then run through these potentially long `for` loops.